### PR TITLE
build: migrate to pyproject.toml

### PR DESCRIPTION
**Pull request by lengau**
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

This removes setup.py and setup.cfg in favour of the newer pyproject.toml file.

It is a necessary portion of our move to starflow workflows, but standalone and per-project.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,177 @@
...
+
+[build-system]
+requires = [
+    "setuptools>=69.0",
+    "setuptools_scm[toml]>=7.1"
+]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools.package-dir]
+"craft_parts_docs" = "docs/common"
```
Comment (sergiusens) [inline pyproject.toml:152]: when adding this it seems it disables automatic package discovery, which is the root cause to the import errors for craft parts itself

```suggestion
[tool.setuptools.package-dir]
"craft_parts" = "craft_parts"
"craft_parts_docs" = "docs/common"
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,178 @@
...
+[tool.setuptools.package-dir]
+"craft_parts" = "craft_parts"
+"craft_parts_docs" = "docs/common"
+
+
+[tool.setuptools.dynamic]
+readme = {file = "README.rst"}
+
+[tool.setuptools_scm]
+write_to = "craft_parts/_version.py"
```
Comment (tigarmo) [inline pyproject.toml:160]: Going to need to update [this](https://github.com/canonical/craft-parts/blob/main/craft_parts/__init__.py#L19)
and remove .bumpversion.cfg

**Review by tigarmo — CHANGES_REQUESTED** [review]

**Review by sergiusens — APPROVED** [review]
considering Tiago's request I consider this good to go

**Comment by lengau** [comment]
Rebased on main, also added bumpversion change.

File: pyproject.toml (hunk 2 above)
Comment (lengau) [reply pyproject.toml:160]: Done! https://github.com/canonical/craft-parts/pull/902/commits/070b8341765435aeb0ceca461b390f4a9e172291

**Comment by lengau** [comment]
Linters and tests are succeeding locally for me - will move out of draft if they succeed on CI.

**Review by mr-cal — APPROVED** [review]

**Review by tigarmo — APPROVED** [review]
thanks
I think it's a good idea to remove the bumpversion.cfg file, as now it's broken and somebody might end up using it by accident

**Comment by lengau** [comment]
@tigarmo I agree - I'd just missed it in my work, so thanks for catching it!

**Pull request merged**